### PR TITLE
Attack bundling

### PR DESCRIPTION
**Pull request by goodfeli**

### CODE REVIEW

File: cleverhans/confidence_report.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+  factory.kwargs['test_start'] = test_start
+  factory.kwargs['test_end'] = test_end
+  dataset = factory()
+
+  center = factory.kwargs['center']
+  value_range = 1. + center
+  min_value = 0. - center
+  max_value = 1.
+
+  if 'CIFAR' in str(factory.cls):
```
Comment (npapernot) [inline cleverhans/confidence_report.py]: Would it make sense to pass these as arguments rather than hardcoding them here so this can be used with other datasets?

File: cleverhans/evaluation.py [hunk 2]
```diff
@@ -100,6 +100,40 @@ def correctness_and_confidence(sess, model, x, y, batch_size=None,
 
   return out
 
+def run_attack(sess, model, x, y, attack, attack_params, batch_size=None,
+               devices=None, feed=None):
+  """
+  Run attack on every example in a dataset.
+  :param sess: tf.Session
+  :param model: cleverhans.model.Model
+  :param x: numpy array containing input examples (e.g. MNIST().x_test )
```
Comment (npapernot) [inline cleverhans/evaluation.py]: See comments above on naming x and y arguments

File: cleverhans/evaluation.py [hunk 3]
```diff
@@ -448,6 +482,49 @@ def __call__(self):
     return (x_batch, y_batch), (correct, max_probs)
 
 
+class _AttackFactory(object):
+  """
+  A factory for an expression that runs an adversarial attack
+  """
+
+  def __init__(self, model, attack=None, attack_params=None, pass_y=False):
```
Comment (npapernot) [inline cleverhans/evaluation.py]: Can you add a comment for `pass_y`? It's possible to find out what it does from line 520 (pass the label placeholder to the attack) but it's not obvious when looking at init

File: scripts/make_confidence_report_bundled.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
+"""
+make_confidence_report_bundled.py
+Usage:
+  python make_confidence_report_bundled.py model.joblib
+
+  where model.joblib is a file created by cleverhans.serial.save containing
+  a picklable cleverhans.model.Model instance.
+
+This script will run the model on clean data and bundled adversarial examples
```
Comment (npapernot) [inline scripts/make_confidence_report_bundled.py]: I don't mind having scripts but could you explain how we should be deciding whether something is a tutorial or a script? Is a script a short example that we want to maintain?

File: cleverhans/attack_bundling.py [hunk 5]
```diff
@@ -0,0 +1,797 @@
...
+num_devices = len(devices)
+DEFAULT_EXAMPLES_PER_DEVICE = 128
+BATCH_SIZE = DEFAULT_EXAMPLES_PER_DEVICE * num_devices
+REPORT_TIME_INTERVAL = 60
+
+# TODO: lower priority: make it possible to initialize one attack with
+# the output of an earlier attack
+
+
+def cheap_max_confidence_recipe(sess, model, x, y, nb_classes, eps,
```
[the remark this answers is not in the document]
Comment (goodfeli) [reply cleverhans/attack_bundling.py]: Changed to `single_run_max_confidence_recipe`

File: cleverhans/attack_bundling.py [hunk 6]
```diff
@@ -0,0 +1,797 @@
...
+                                clip_min, clip_max, eps_iter, nb_iter,
+                                report_path,
+                                batch_size=BATCH_SIZE):
+  """A reasonable attack bundling recipe for a max norm threat model and
+  a defender that uses confidence thresholding.
+
+  References:
+  https://openreview.net/forum?id=H1g0piA9tQ
+
+  This version is "cheap" because it runs each attack just once and
```
[the remark this answers is not in the document]
Comment (goodfeli) [reply cleverhans/attack_bundling.py]: Done

File: cleverhans/attack_bundling.py [hunk 7]
```diff
@@ -0,0 +1,797 @@
...
+  indefinitely.
+
+  :param sess: tf.Session
+  :param model: cleverhans.model.Model
+  :param x: numpy array containing clean example inputs to attack
+  :param y: numpy array containing true labels
+  :param nb_classes: int, number of classes
+  :param eps: float, maximum size of perturbation (measured by max norm)
+  :param eps_iter: float, step size for cheaper PGD attacks
+    (will also run more expensive ones with 25X smaller step size)
```
[the remark this answers is not in the document]
Comment (goodfeli) [reply cleverhans/attack_bundling.py]: I just changed it to say there are 2 versions and one has 25X more iterations. I don't use an adjective like "cheap" to describe the first one or "expensive" to describe the second one anymore.

File: cleverhans/attack_bundling.py [hunk 8]
```diff
@@ -0,0 +1,797 @@
...
+
+  This version runs indefinitely, updating the report on disk continuously.
+
+  :param sess: tf.Session
+  :param model: cleverhans.model.Model
+  :param x: numpy array containing clean example inputs to attack
+  :param y: numpy array containing true labels
+  :param nb_classes: int, number of classes
+  :param eps: float, maximum size of perturbation (measured by max norm)
+  :param eps_iter: float, step size for cheaper PGD attacks
```
[the remark this answers is not in the document]
Comment (goodfeli) [reply cleverhans/attack_bundling.py]: done

File: cleverhans/attack_bundling.py [hunk 9]
```diff
@@ -0,0 +1,797 @@
...
+    (will also run more expensive ones with 25X smaller step size)
+  :param nb_iter: int, number of iterations for the cheaper PGD attacks
+    (will also run more expensive ones with 25X more iterations)
+  :param report_path: str, the path that the report will be saved to.
+  :batch_size: int, the total number of examples to run simultaneously
+  """
+  noise_attack = Noise(model, sess)
+  pgd_attack = ProjectedGradientDescent(model, sess)
```
[the remark this answers is not in the document]
Comment (goodfeli) [reply cleverhans/attack_bundling.py]: I think it would hurt readability.
I intend for people to be able to copy-paste and modify a recipe easily without needing to understand several layers of abstraction.

File: cleverhans/attack_bundling.py [hunk 10]
```diff
@@ -0,0 +1,797 @@
...
+  :param model: cleverhans.model.Model
+  :param x: numpy array containing clean example inputs to attack
+  :param y: numpy array containing true labels
+  :param adv_x: numpy array containing the adversarial examples made so far
+    by earlier work in the bundling process
+  :param attack_configs: list of AttackConfigs to run
+  :param run_counts: dict mapping AttackConfigs to numpy arrays specifying
+    how many times they have been run on each example
+  :param goal: AttackGoal to run
+  :param report: dict
```
[the remark this answers is not in the document]
Comment (goodfeli) [reply cleverhans/attack_bundling.py]: I think I'd like to postpone that to a later PR, but do a more comprehensive solution based on a ConfidenceReport class: https://github.com/tensorflow/cleverhans/issues/768

File: cleverhans/attack_bundling.py [hunk 11]
```diff
@@ -0,0 +1,797 @@
...
+  """
+  Runs attack bundling on one batch of data.
+  This function is mostly intended to be called by
+  `bundle_attacks_with_goal`.
+
+  :param sess: tf.session.Session
+  :param model: cleverhans.model.Model
+  :param x: numpy array containing clean example inputs to attack
+  :param y: numpy array containing true labels
+  :param adv_x: numpy array containing the adversarial examples made so far
```
[the remark this answers is not in the document]
Comment (goodfeli) [reply cleverhans/attack_bundling.py]: done

File: cleverhans/attack_bundling.py [hunk 12]
```diff
@@ -0,0 +1,797 @@
...
+    In the future we may support other schemes like smaller perturbation
+    size, higher loss, etc.
+  """
+
+  def __init__(self, new_work_goal=None, break_ties='wrong_confidence'):
+    super(Misclassify, self).__init__()
+    self.new_work_goal = new_work_goal
+    assert all(isinstance(key, AttackConfig) for key in new_work_goal.keys())
+    assert all(isinstance(value, int) for value in new_work_goal.values())
+    self.rng = np.random.RandomState([2018, 10, 5, 9])
```
[the remark this answers is not in the document]
Comment (goodfeli) [reply cleverhans/attack_bundling.py]: We can make an issue but I'm not sure what the concern is. This RNG here is not particularly important. It is just used to run the attacks in a random order. Everything would still work if the attacks were run in a fixed order. The random order is just easier to implement because it doesn't require keeping track of interator state between calls to get_attack_config.

File: cleverhans/confidence_report.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+TRAIN_END = 60000
+TEST_START = 0
+TEST_END = 10000
+WHICH_SET = 'test'
+
+def make_confidence_report_bundled(filepath, train_start=TRAIN_START,
+                                   train_end=TRAIN_END, test_start=TEST_START,
+                                   test_end=TEST_END, which_set=WHICH_SET):
+  """
+  Load a saved model, gather its predictions, and save a confidence report.
```
[the remark this answers is not in the document]
Comment (goodfeli) [reply cleverhans/confidence_report.py]: We would need that kind of check if we were writing a function where someone passes us a Model object reference and later we plan to call `save` on it. Here they pass us a filepath and we call `load` on the filepath. Because the model is already in a .joblib file we know it was picklable.

File: cleverhans/confidence_report.py [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+  :param train_end: index of last training set example to use
+  :param test_start: index of first test set example to use
+  :param test_end: index of last test set example to use
+  :param which_set: 'train' or 'test'
+  """
+  # Avoid circular import
+  from cleverhans.attack_bundling import basic_max_confidence_recipe
+
+  # Set TF random seed to improve reproducibility
+  tf.set_random_seed(1234)
```
[the remark this answers is not in the document]
Comment (goodfeli) [reply cleverhans/confidence_report.py]: yeah, that one I should remove, because it does affect the attacks. Took it out.

File: cleverhans/confidence_report.py (hunk 1 above)
Comment (goodfeli) [reply cleverhans/confidence_report.py]: At the moment, these really are just hard-coded values that people use for each dataset in published papers. I don't know a way of auto-inferring them for new datasets.

File: cleverhans/evaluation.py (hunk 2 above)
Comment (goodfeli) [reply cleverhans/evaluation.py]: These match the other functions in the file

File: cleverhans/evaluation.py (hunk 3 above)
Comment (goodfeli) [reply cleverhans/evaluation.py]: done

File: scripts/make_confidence_report_bundled.py (hunk 4 above)
Comment (goodfeli) [reply scripts/make_confidence_report_bundled.py]: A script is a tool that we'll use frequently while actually doing work. A tutorial is something that is written for the purpose of showing how the library works, but is overly simplified and not practical for actual use (e.g. it builds a really small model).

File: cleverhans/confidence_report.py (hunk 1 above)
Comment (goodfeli) [reply cleverhans/confidence_report.py]: Later we might add some kind of config system. I'd prefer to save that for a later PR though.

File: cleverhans/attack_bundling.py [hunk 15]
```diff
@@ -0,0 +1,797 @@
...
+
+# TODO: lower priority: make it possible to initialize one attack with
+# the output of an earlier attack
+
+
+def cheap_max_confidence_recipe(sess, model, x, y, nb_classes, eps,
+                                clip_min, clip_max, eps_iter, nb_iter,
+                                report_path,
+                                batch_size=BATCH_SIZE):
+  """A reasonable attack bundling recipe for a max norm threat model and
```
[the remark this answers is not in the document]
Comment (goodfeli) [reply cleverhans/attack_bundling.py]: done

File: cleverhans/attack_bundling.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (npapernot) [reply cleverhans/attack_bundling.py]: Thanks, the new version is much more clear about the difference between both attacks

File: cleverhans/attack_bundling.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (npapernot) [reply cleverhans/attack_bundling.py]: Sounds good, I think we should keep it as is if that's the intended use case

File: cleverhans/attack_bundling.py [hunk 16]
```diff
@@ -0,0 +1,797 @@
...
+  """
+  Runs attack bundling.
+  Users of cleverhans may call this function but are more likely to call
+  one of the recipes above.
+
+  Reference: https://openreview.net/forum?id=H1g0piA9tQ
+
+  :param sess: tf.session.Session
+  :param model: cleverhans.model.Model
+  :param x: numpy array containing clean example inputs to attack
```
[the remark this answers is not in the document]
Comment (npapernot) [reply cleverhans/attack_bundling.py]: Given that we have to create the placeholders within the train and evaluate functions of the library to support multi-GPU, I agree it makes sense to reflect that change in the new API. I don't understand though why not keep a name (e.g., x_train and y_train) that makes it clear to users that they should pass numpy arrays and not tensors? If I remember correctly, I think that's what train and batch_eval do in their multi GPU variants. I'm sorry I must have missed this when merging `correctness_and_confidence`. I'm just worried that users will get confused if they miss the change, and having this name convention there would not change the implementation, it just makes it easier for users to notice the change? Maybe I'm missing something?

Regarding the second part of your comment, I'm not sure we want to completely remove the ability to use placeholders directly with the attack classes. I feel like it would hurt flexibility. For the rest of the library (training, eval), I understand why that makes sense and agree we should keep going in that direction. We should probably have this discussion with tf.eager in mind too since it will likely play in the design as well.

File: cleverhans/attack_bundling.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (npapernot) [reply cleverhans/attack_bundling.py]: Sounds good!

File: cleverhans/attack_bundling.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (npapernot) [reply cleverhans/attack_bundling.py]: Ok I see, then there is no need to make an issue for this then.

File: cleverhans/confidence_report.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (npapernot) [reply cleverhans/confidence_report.py]: Thanks for clarifying, I agree with you

File: cleverhans/confidence_report.py (hunk 1 above)
Comment (npapernot) [reply cleverhans/confidence_report.py]: Ah I see why the values were hard-coded then. Would it make sense to move this to scripts or do you think it's a feature we want people to be able to import directly from the CleverHans module?

File: scripts/make_confidence_report_bundled.py (hunk 4 above)
Comment (npapernot) [reply scripts/make_confidence_report_bundled.py]: Sounds good, I'll make an issue to clarify this in the README file for future reference

Comment (goodfeli) [reply scripts/make_confidence_report_bundled.py]: SGTM

File: cleverhans/attack_bundling.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (goodfeli) [reply cleverhans/attack_bundling.py]: We don't want to use a name like `x_train` here because this could also be the validation or test set. Since this code is used for benchmarking, it's usually going to be the test set.

A lot of our other naming conventions haven't worked out very well. We've had a lot of "_ph" names for variables that don't actually need to be placeholders. We're also not very consistent with "_val" and "_value". Sometimes "_value" means that a variable is a tensorflow tensor rather than a callable, sometimes "_value" means that a variable is numpy rather than tensorflow. So I find the names are usually just no typing but don't add any information, and sometimes mislead me.

How about using `x` for the name and, in addition to the docstring saying `x` is numpy, also have an if statement that raises a ValueError immediately if `x` is not numpy?

It's still possible for users to make placeholders, but pretty soon the default will be for them to not do that. Also, as time goes on, it will be an increasingly dumb thing for them to do. Already, making your own placeholder is a pretty bad decision, because it means everything you write after that will need to be rewritten if you ever want to upgrade to using the multi-GPU features. So while we still want to allow it, we should give people strong nudges away from doing it.

**Comment by goodfeli** [comment]
I added type checks on `x` and `y` to the methods in `cleverhans.evaluation` that take `x` and `y` arguments. The type check functions will immediately raise a `TypeErorr` with an explanation of what `x` and `y` should be if the user passes an invalid value.
